src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php

```diff
@@ -21,18 +25,69 @@
  *   }
  * )
  */
-class ImageUrl extends DataProducerPluginBase {
+class ImageUrl extends DataProducerPluginBase implements ContainerFactoryPluginInterface {
 
   /**
-   * @param \Drupal\Core\Entity\EntityInterface $entity
+   * The renderer service.
+   *
+   * @var \Drupal\Core\Render\RendererInterface
+   */
+  protected $renderer;
+
+  /**
+   * {@inheritdoc}
+   *
+   * @codeCoverageIgnore
+   */
+  public static function create(ContainerInterface $container, array $configuration, $pluginId, $pluginDefinition) {
+    return new static(
+      $configuration,
+      $pluginId,
+      $pluginDefinition,
+      $container->get('renderer')
+    );
+  }
+
+  /**
+   * EntityRendered constructor.
+   *
+   * @param array $configuration
+   *   The plugin configuration array.
+   * @param string $pluginId
+   *   The plugin id.
+   * @param mixed $pluginDefinition
+   *   The plugin definition.
+   * @param \Drupal\Core\Render\RendererInterface $renderer
+   *   The renderer service.
+   *
+   * @codeCoverageIgnore
+   */
+  public function __construct(
+    array $configuration,
+    $pluginId,
+    $pluginDefinition,
+    RendererInterface $renderer
+  ) {
+    parent::__construct($configuration, $pluginId, $pluginDefinition);
+    $this->renderer = $renderer;
+  }
+
+  /**
+   * @param \Drupal\file\FileInterface $entity
+   *   The file entity.
    *
    * @return mixed
+   *   The file url.
    */
   public function resolve(FileInterface $entity, RefinableCacheableDependencyInterface $metadata) {
     $access = $entity->access('view', NULL, TRUE);
     $metadata->addCacheableDependency($access);
     if ($access->isAllowed()) {
-      return file_create_url($entity->getFileUri());
+      $context = new RenderContext();
+      $file_url = $this->renderer->executeInRenderContext($context, function () use ($entity, $image_style, $dimensions) {
+        return file_create_url($entity->getFileUri());
+      });
+      return $file_url;
```

Comment:
this variable could be uninitialized here if access is denied. So we should set it first to NULL?

Reply:
agree yes I think we should set it to null.if it fails I'd be ok with returning null